src/gc/unix/cgroup.cpp

```diff
@@ -118,17 +118,10 @@ class CGroup
             *val = 1;
             return true;
         }
-        
-        cpu_count = (double) quota / period;
-        if (cpu_count < UINT32_MAX - 1)
-        {
-            // round up
-            *val = (uint32_t)(cpu_count + 0.999999999);
-        }
-        else
-        {
-            *val = UINT32_MAX;
-        }
+
+        // Calculate cpu count based on quota and round it up
+        cpu_count = (double) quota / period  + 0.999999999;
```

Comment:
quota / period == quota * reciprocal of the period as a double and would be faster depending on compiler flags.

Reply:
Multiplication by the reciprocal is not an equivalent operation for floating-point values.

Reply:
as we casted to double here does that matter?

Reply:
https://cs5thmatho2011.wordpress.com/2011/10/23/using-reciprocals-to-turn-division-into-multiplication/

Reply:
It can. Right now the code is performing four operations.

1. Conversion of `quota` to a double
2. Conversion of `period` to a double
3. Dividing the result of 1 by the result of 2
4. Adding 0.999999999 to the result of 3

If you change it to multiply be the reciprocal, you will be adding an additional step
1. Conversion of `quota` to a double
2. Conversion of `period` to a double
3. Division of 1.0 by the result of 2
4. Multiplication of the result of 1 by the result of 3
5. Adding 0.999999999 to the result of 4

Each floating-point operation takes the given inputs as internally represented (using all 64-bits for double), computes the result to "infinite" precision, and then rounds to the nearest representable result. This means that every result can have some level of rounding error and the more steps you do, the greater that rounding error can be.

This is also why something like: `(5.0 * (1.0 / 1000000.0)) == 4.9999999999999996E-06` but `(5.0 / 1000000.0) == 5E-06`

Reply:
But division is multiple operations, I know its one instruction but the latency vs mul would easily trump it right?

Reply:
> But division is multiple operations in the pipeline, I know its one instruction but the latency vs mul would easily trump it right?

You have a division here either way (in getting the reciprocal); so the multiplication by reciprocal trick is really only useful when you are doing multiple operations and you are fine with losing some precision.

You are also thinking of integer division. Agner's shows that floating-point multiplication and division are both one micro-op and have relatively comparable latency (2-4x slower); as compared to integer multiplication and division which can take 1-4 micro ops for mul and 10-57 micro ops for division and have largely different latencies (6-10x slower, or more for division)